File: core/src/commonMain/kotlin/io/silencelen/andvari/core/crypto/Bytes.kt
package io.silencelen.andvari.core.crypto

import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

/**
 * Spec 00 conventions: binary ⇄ text is base64url without padding, everywhere;
 * hex only for printed fingerprints.
 */
@OptIn(ExperimentalEncodingApi::class)
object Bytes {
    private val b64 = Base64.UrlSafe.withPadding(Base64.PaddingOption.ABSENT_OPTIONAL)

    fun toB64(bytes: ByteArray): String = b64.encode(bytes)

    fun fromB64(text: String): ByteArray = try {
        b64.decode(text)
    } catch (e: IllegalArgumentException) {
        throw CryptoException("invalid base64url", e)
    }

    @OptIn(ExperimentalStdlibApi::class)
    fun toHexLower(bytes: ByteArray): String = bytes.toHexString()

    @OptIn(ExperimentalStdlibApi::class)
    fun toHexUpper(bytes: ByteArray): String = bytes.toHexString(HexFormat.UpperCase)

    /** Constant-time comparison for MAC-like values. */
    fun ctEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var acc = 0
        for (i in a.indices) acc = acc or (a[i].toInt() xor b[i].toInt())
        return acc == 0
    }

    /**
     * Shape (the first) 16 bytes into an RFC 4122 v4-format UUID string (version nibble
     * forced to 4, variant to 10) — the ONE shared implementation behind random ids
     * (Account.uuid), deterministic conflict-copy ids (ConflictCopy, vector-pinned), and
     * F19 gesture mutationIds (SyncEngine), byte-identical to web's uuidFromLabel shaping.
     * A drift between copies silently breaks server-dedup convergence, so everything
     * derives from this.
     */
    fun uuidV4FromBytes(bytes: ByteArray): String {
        require(bytes.size >= 16) { "need at least 16 bytes for a UUID" }
        val h = bytes.copyOf(16)
        h[6] = ((h[6].toInt() and 0x0F) or 0x40).toByte() // version nibble 4
        h[8] = ((h[8].toInt() and 0x3F) or 0x80).toByte() // variant 10
        val hex = toHexLower(h)
        return "${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20)}"
    }
}

File: web/src/crypto/bytes.ts
import { CryptoError, sodium } from "./sodium";

/** Spec 00 conventions: base64url unpadded everywhere; hex only for fingerprints. */

export function toB64(bytes: Uint8Array): string {
  const s = sodium();
  return s.to_base64(bytes, s.base64_variants.URLSAFE_NO_PADDING);
}

export function fromB64(text: string): Uint8Array {
  const s = sodium();
  try {
    return s.from_base64(text, s.base64_variants.URLSAFE_NO_PADDING);
  } catch (e) {
    throw new CryptoError("invalid base64url", { cause: e });
  }
}

export function utf8(text: string): Uint8Array {
  return new TextEncoder().encode(text);
}

export function fromUtf8(bytes: Uint8Array): string {
  return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
}

export function toHexLower(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

export function toHexUpper(bytes: Uint8Array): string {
  return toHexLower(bytes).toUpperCase();
}

export function concat(...parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

/** Constant-time comparison for MAC-like values. */
export function ctEquals(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let acc = 0;
  for (let i = 0; i < a.length; i++) acc |= (a[i]! ^ b[i]!);
  return acc === 0;
}
